### PR TITLE
🧑‍🔬 Prototype: Enforcing MFA for owners of top 100 downloaded gems (UI)(Option 1️⃣)

### PR DESCRIPTION
**Pull request by bettymakes**
Contributes to https://github.com/Shopify/code-foundations/issues/166

## 🛠️ Prototypes
We spiked 3 different approaches for implementing _enforcing MFA for owners of top 100 downloaded gems_:

**Option 1️⃣ PR: [Programmatic - dynamically determine owners of top 100 gems](https://github.com/Shopify/rubygems.org/pull/2)** 
☝️ This is the current PR. In this implementation, we're dynamically determining whether a user should have the MFA requirement enforced on their accounts. The risk to this approach is that MFA requirements may change day-to-day (or even minute-by-minute). For example, it's possible for gems along the edge of the "top 100 gems" threshold (e.g. gem ranked 99, 100, 101, 102 etc.) to change positions. This presents an inconsistency for whether certain accounts require mfa or not.  

Option 2️⃣ PR: [Add database column (boolean) - identify mfa-required users by mfa_required field](https://github.com/Shopify/rubygems.org/pull/6)

Option 3️⃣ PR: [Add database column (date) - provide an mfa enrolment deadline, upon deadline mfa enforcement becomes a requirement](https://github.com/Shopify/rubygems.org/pull/8)

## 🧪 What problem does this prototype solve?
Please refer to 

**TL;DR – This prototype requires owners of the top 100 downloaded gems to enable MFA.** Once enabled, those users will not be able to disable MFA. They can however, deregister a device and register with a new one. However, they will not be able to access any other views on `rubygems.org` that require authentication until MFA is enabled. 

---

Account takeovers is second most common vulnerability that's exploited. In enforcing MFA, we add more friction and increase complexity of successfully exploiting that vulnerability.

In an ideal world, all accounts registered on `rubygems.org` require MFA. Recognizing that policy may require a lot more change management efforts for all stakeholders to be bought-in, we brainstormed options that'll get us from 1 to 2, as opposed to 1 to 10. 

And so, instead of enforcing MFA on _everyone_, an incremental approach could be to target the highest-risk accounts. We identify higher risk accounts as `accounts belonging to users who are owners of the top 100 most downloaded gems`, since the blast radius of a nefarious actor taking over high risk accounts is quite large. 

## 🧬 How does this prototype work?
This introduces 3 different distinct user flows:

🔹 **When a user is an owner of a top 100 gem, with MFA disabled**
Once the user signs in to `rubygems.org`, they are redirected to the MFA set up view. All views that require authentication will not be accessible to the user until MFA is set up. They will still be able to view any publicly accessible views that do not require authentication.

The option to disable MFA is removed for these users. However, if they want to switch their authentication device, they can deregister the current device, and register a new one. Once their current device is deregistered, they cannot access any views that require authentication until they register their new device.

⚠️ The risk here is that it's possible for a user to deregister their device, technically disabling MFA. And, if they do not follow through with the second step of registering a new device, their account is once again _more vulnerable_ to account takeover attacks.

🎬 Demo: User flow of a top gem owner without MFA enabled

https://user-images.githubusercontent.com/4270287/149182021-376e44cb-b484-4ee9-96fe-aa1446659d0c.mov

🎬 Demo: User flow of a top gem owner deregistering their device (thereby disabling MFA)

https://user-images.githubusercontent.com/4270287/149188379-2a2602ce-de60-400c-9ddc-696f4e8c8137.mov

🔹 **When a user is an owner of a top 100 gem, with MFA enabled**
This is the happy path! Once the user signs in to `ruybgems.org`, they can access all the views.

🎬 Demo: User flow of a top gem owner setting up MFA

https://user-images.githubusercontent.com/4270287/149187258-67e9be42-495f-41f8-bb50-e44e5e004063.mov

🎬 Demo: User flow of a top gem owner signing in with MFA already enabled

https://user-images.githubusercontent.com/4270287/149187236-eb39b4ed-bd60-427e-bd3a-e91a143900a2.mov

🔹 **When a user is not an owner of a top 100 gem**
There are no changes to this user's experience on `rubygems.org`, as they are not in the cohort this prototype is proposing to enforce MFA on.

🎬 Demo: User flow for users who are not owners of a top 100 gem

https://user-images.githubusercontent.com/4270287/149189785-949d1f37-8f40-4966-8d33-45cd46ee6aba.mov

## 📝 Other notes
* This prototype only focuses on the UI experience and not CLI.

### CODE REVIEW

File: app/models/user.rb [hunk 1]
```diff
@@ -228,6 +228,18 @@ def mfa_api_authorized?(otp)
...
+  def mfa_required?
+    return false if mfa_enabled?
+    owner_of_most_downloaded_gem?
+  end
+
+  # NOTE:
+  # Do we want to differentiate the concept of mfa required for an acct where mfa is not enabled vs.
+  # Identifying which accounts require mfa, regardless whether it's enabled or not
+  def owner_of_most_downloaded_gem?
+    (rubygems & Rubygem.by_downloads.limit(Rubygem::MOST_DOWNLOADED)).any?
```
Comment (doodzik) [inline app/models/user.rb:240]: I'm having a hard time wrapping my head around this line. It is probably good to add a code comment to explain it a bit.

**Comment by doodzik** [comment]
I like the proposal and makes sense 👍 

> ⚠️ The risk here is that it's possible for a user to deregister their device, technically disabling MFA. And, if they do not follow through with the second step of registering a new device, their account is once again more vulnerable to account takeover attacks.

I wonder if you could move the removal and register of a new device into a single step? So you could avoid this situation.

A related question, I was wondering about an attack where someone gets hold of someone's account by obtaining the MFA access or taking control of a running session. Have you thought of blocking someone from making changes for a while when deregistering an account?  That way there is the possibility of reverting the change by the user or notifying people that the account may be compromised.

File: app/models/user.rb (hunk 1 above)
Comment (jenshenny) [inline app/models/user.rb:240]: Ashley and I had a quick convo about how we want the determine who is required to enable MFA. Right now it is dynamic, this raises some benefits and concerns:

- What if gem 101 becomes 100 when we release this policy? It's good that now they are required to enable MFA but if they become 100, they don't have any transition time when this hits. gem 100 becomes gem 101 and now they can disable MFA which is also weird.
- The query time of this method could slow these requests down a bit.

Here are the pros and cons for a static list:
- Can't enforce gems that surge in popularity during the policy period
- Is it PII to commit a list of gems? How are we storing this list?
- Faster to check a static list
- We know what gem accounts we are targeting and there are no surprises

With the reasons above, I'm leaning towards the static list.

File: app/views/settings/edit.html.erb [hunk 2]
```diff
@@ -1,21 +1,38 @@
...
-        [t(".mfa.level.ui_and_gem_signin"), "ui_and_gem_signin"],
-        [t(".mfa.level.ui_only"), "ui_only"],
-        [t(".mfa.level.disabled"), "disabled"]], @user.mfa_level), class: "form__input form__select" %>
+      <%# Note: Scrappy version ... this can obviously be DRY'd up %>
+      <% if @user.owner_of_most_downloaded_gem? %>
+         <%= select_tag :level, options_for_select([
+          [t(".mfa.level.ui_and_api"), "ui_and_api"],
+          [t(".mfa.level.ui_and_gem_signin"), "ui_and_gem_signin"],
+          [t(".mfa.level.ui_only"), "ui_only"],
+          ["Deregister device", "disabled"]], @user.mfa_level), class: "form__input form__select" %>
```
Comment (jenshenny) [inline app/views/settings/edit.html.erb:14]: It's great to see a different drop down with the available options `ui and api`, `ui and gem signin` and `deregister a device`! I think we should also add some validation in the backend to make sure nobody is trying to set their level to `ui_only` by modifying the DOM.

**Comment by jenshenny** [comment]
> A related question, I was wondering about an attack where someone gets hold of someone's account by obtaining the MFA access or taking control of a running session. Have you thought of blocking someone from making changes for a while when deregistering an account? That way there is the possibility of reverting the change by the user or notifying people that the account may be compromised.

If I understand correctly, I don't think I see the purpose of blocking changes for a while. For context, you need to enter an otp code when changing your MFA level. If a user already has access to the compromised user's MFA, they do not need to deregister the MFA device to make harmful changes.

File: app/views/settings/edit.html.erb (hunk 2 above)
Comment (aellispierce) [reply app/views/settings/edit.html.erb]: Addressed in https://github.com/Shopify/rubygems.org/pull/2/commits/394bf77f7672e42b3df73776b165f2c5e9d6bc52

**Comment by bettymakes** [comment]
Closing in favour of https://github.com/Shopify/rubygems.org/pull/10 since it's the amalgamation of all the prototypes.

**Pull request closed without merging**